### PR TITLE
Prevent multiple calls to ExtractDestination

### PR DESCRIPTION
**Pull request by pedrobranco**
Prevent multiple calls to ExtractDestination in RPC call `listunspent`.

@promag @jpdffonseca

### CODE REVIEW

**Comment by joaopaulofonseca** [comment]
Nice small code cleanup.

ACK

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2402,28 +2405,25 @@ UniValue listunspent(const UniValue& params, bool fHelp)
...
         entry.push_back(Pair("vout", out.i));
-        CTxDestination address;
-        if (ExtractDestination(out.tx->vout[out.i].scriptPubKey, address)) {
+        if (bExtractResult) {
             entry.push_back(Pair("address", CBitcoinAddress(address).ToString()));
             if (pwalletMain->mapAddressBook.count(address))
                 entry.push_back(Pair("account", pwalletMain->mapAddressBook[address].name));
         }
         entry.push_back(Pair("scriptPubKey", HexStr(pk.begin(), pk.end())));
         if (pk.IsPayToScriptHash()) {
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: `if (bExtractResult && pk.IsPayToScriptHash()) {` and remove the ttest below.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -2392,9 +2392,12 @@ UniValue listunspent(const UniValue& params, bool fHelp)
         if (out.nDepth < nMinDepth || out.nDepth > nMaxDepth)
             continue;
 
+        CTxDestination address;
+        const CScript& pk = out.tx->vout[out.i].scriptPubKey;
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: rename pk to scriptPubKey.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -2392,9 +2392,12 @@ UniValue listunspent(const UniValue& params, bool fHelp)
...
+        bool bExtractResult = ExtractDestination(pk, address);
+
         if (setAddress.size()) {
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: `if (setAddress.size() && !boost::get<CNoDestination>(&address)) {` and  remove the condition below. But verify if this is correct.

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -2392,9 +2392,12 @@ UniValue listunspent(const UniValue& params, bool fHelp)
         if (out.nDepth < nMinDepth || out.nDepth > nMaxDepth)
             continue;
 
+        CTxDestination address;
+        const CScript& pk = out.tx->vout[out.i].scriptPubKey;
+        bool bExtractResult = ExtractDestination(pk, address);
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Don't save bExtractResult.

**Comment by pedrobranco** [comment]
Fixed and rebased to master.

**Comment by pedrobranco** [comment]
@promag @jpdffonseca review please

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -2392,38 +2392,33 @@ UniValue listunspent(const UniValue& params, bool fHelp)
...
-            if (!ExtractDestination(out.tx->vout[out.i].scriptPubKey, address))
-                continue;
+        CTxDestination address;
+        const CScript& scriptPubKey = out.tx->vout[out.i].scriptPubKey;
+        ExtractDestination(scriptPubKey, address);
 
-            if (!setAddress.count(address))
-                continue;
-        }
+        if (setAddress.size() && ( boost::get<CNoDestination>(&address) || !setAddress.count(address) ) )
```
Comment (joaopaulofonseca) [inline src/wallet/rpcwallet.cpp]: Remove extra spaces inside parenthesis

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -2392,38 +2392,33 @@ UniValue listunspent(const UniValue& params, bool fHelp)
...
+            continue;
 
         CAmount nValue = out.tx->vout[out.i].nValue;
-        const CScript& pk = out.tx->vout[out.i].scriptPubKey;
         UniValue entry(UniValue::VOBJ);
         entry.push_back(Pair("txid", out.tx->GetHash().GetHex()));
         entry.push_back(Pair("vout", out.i));
-        CTxDestination address;
-        if (ExtractDestination(out.tx->vout[out.i].scriptPubKey, address)) {
+        if (!boost::get<CNoDestination>(&address)){
```
Comment (joaopaulofonseca) [inline src/wallet/rpcwallet.cpp]: Add space before last brace

**Comment by pedrobranco** [comment]
Format done.

**Comment by pedrobranco** [comment]
@promag Ready to go upstream?

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -2390,42 +2390,37 @@ UniValue listunspent(const UniValue& params, bool fHelp)
...
+        if (!boost::get<CNoDestination>(&address) && scriptPubKey.IsPayToScriptHash()) {
+            const CScriptID& hash = boost::get<CScriptID>(address);
+            CScript redeemScript;
+            if (pwalletMain->GetCScript(hash, redeemScript))
+                entry.push_back(Pair("redeemScript", HexStr(redeemScript.begin(), redeemScript.end())));
         }
-        entry.push_back(Pair("amount",ValueFromAmount(nValue)));
-        entry.push_back(Pair("confirmations",out.nDepth));
+
+        entry.push_back(Pair("scriptPubKey", HexStr(scriptPubKey.begin(), scriptPubKey.end())));
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Sort keys.

File: src/wallet/rpcwallet.cpp [hunk 8]
```diff
@@ -2390,42 +2390,37 @@ UniValue listunspent(const UniValue& params, bool fHelp)
     assert(pwalletMain != NULL);
     LOCK2(cs_main, pwalletMain->cs_wallet);
     pwalletMain->AvailableCoins(vecOutputs, false, NULL, true);
-    BOOST_FOREACH(const COutput& out, vecOutputs) {
+    BOOST_FOREACH (const COutput& out, vecOutputs) {
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Remove this change.

**Comment by pedrobranco** [comment]
Deprecated by upstream PR in bitcoin. PR 7825.

**Pull request closed without merging**